optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,371 @@
+//                      Math.abs($leftPeriodIndex - periodIndex) < ($constraintConfiguration.getPeriodSpreadLength() + 1))
+//    then
+//        scoreHolder.penalize(kcontext, $topicConflict.getStudentSize());
+//    end
+    private Constraint periodSpread(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
+                .filter((topicConflict, leftExam, rightExam) -> Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex()) > 1)
+                .penalize("Period spread",
+                          HardSoftScore.ONE_SOFT,
+                          (topicConflict, leftExam, rightExam) -> topicConflict.getStudentSize());
+    }
+
+    //    // Several exams in the same room and period have different durations
+//    rule "mixedDurations"
+//    when
+//    $leftExam : Exam($leftId : id, period != null, $period : period, room != null, $room : room,
+//                     $leftTopicDuration : topicDuration)
+//    // 4 mixed durations of 100, 150, 200 and 200 should only result in 2 penalty's (for 100&150 and 100&200)
+//    // leftExam has lowest id of the period+room combo
+//    not Exam(period == $period, room == $room, id < $leftId)
+//    // rightExam has a different duration
+//    $rightExam : Exam(period == $period, room == $room, id > $leftId,
+//                      topicDuration != $leftTopicDuration,
+//                      $rightId : id, $rightTopicDuration : topicDuration
+//    )
+//    // rightExam has the lowest id of the period+room+rightDuration combo
+//    not Exam(period == $period, room == $room, id < $rightId,
+//            topicDuration == $rightTopicDuration
+//    )
+//    then
+//        scoreHolder.penalize(kcontext);
+//    end
+    private Constraint mixedDurations(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
+                .ifExists(Exam.class, equal(Exam::getRoom, Exam::getRoom))
+                .ifNotExistsOther(Exam.class,
+                                  equal(Exam::getPeriod, Exam::getPeriod),
+                                  equal(Exam::getRoom, Exam::getRoom),
+                                  lessThan(Exam::getId, Exam::getId))
+                .join(Exam.class,
+                      equal(Exam::getPeriod, Exam::getPeriod),
+                      equal(Exam::getRoom, Exam::getRoom),
+                      greaterThan(Exam::getId, Exam::getId))
```

Comment:
Since both left and right are the same type and we're joining on the same field, `equal()` can lose its second argument.